DOC: sparse: Add migration guide for converting code from spmatrix to sparray

dschult opened the pull request:
This Draft PR provides a place to publish and discuss an early draft for sparse migration guidelines from `spmatrix` to `sparray`.

As originally posted, it uses markdown and is not integrated with the html docs. Once the content is mature and vetted, we will convert to rst and integrate with the rest of the docs. 

 EDIT @rgommers: this transition guide is part of a larger plan, see [this mailing list discussion](https://mail.python.org/archives/list/scipy-dev@python.org/message/2CZBCBWP2LMQW67MWKMDJ74PSPGRAXP2/)

dschult committed:
Draft migration guide for converting from spmatrix to sparray [docs only]

dschult committed:
Convert to rst, put in reference doc w/ link from release [docs only]

dschult committed:
reformat tables and misc other changes [docs only]

stefanv commented on doc/source/reference/sparse.migration_to_sparray.rst:
```diff
@@ -0,0 +1,268 @@
+Migration from spmatrix to sparray
+==================================
+
+This document provides guidance for converting code from sparse matrices
+to sparse arrays in SciPy Sparse.
```
```suggestion
to sparse arrays in ``scipy.sparse``.
```

stefanv replied:
```suggestion
This document provides guidance for converting code from sparse *matrices*
to sparse *arrays* in SciPy Sparse.
```

stefanv commented on doc/source/reference/sparse.migration_to_sparray.rst:
```diff
@@ -0,0 +1,268 @@
+
+The change from sparse matrices to sparse arrays mirrors conversion from
+``np.matrix`` to ``np.array``. Essentially we must move from an all-2D
+matrix multiplication centric ``matrix`` object to a 1D or 2D “array”
```
```suggestion
matrix-multiplication-centric ``matrix`` object to a 1D or 2D “array”
```

stefanv commented on doc/source/reference/sparse.migration_to_sparray.rst:
```diff
@@ -0,0 +1,268 @@
+The change from sparse matrices to sparse arrays mirrors conversion from
+``np.matrix`` to ``np.array``. Essentially we must move from an all-2D
+matrix multiplication centric ``matrix`` object to a 1D or 2D “array”
+object that supports operand matrix multiplication and elementwise
```
```suggestion
object that supports the matrix multiplication operator and elementwise
```

stefanv commented on doc/source/reference/sparse.migration_to_sparray.rst:
```diff
@@ -0,0 +1,268 @@
+-  the constructor names ``*_matrix``, e.g. ``csr_matrix``, are changed
+   to ``*_array``.
+-  spmatrix ``M`` is always 2D (rows x columns) even
+   e.g. ``M.min(axis=0)``. sparray ``A`` can be 1D or 2D. Numpy scalars
```
Perhaps missing something like:

> sparray preserves NumPy semantics, so that ``M.min(axis=0)`` returns a 1D result.

stefanv replied:
Not sure what it means to say "Numpy scalars are used as 0D values where needed."

dschult replied:
I included an example to make it more clear:
```
Numpy scalars are returned for full reductions, i.e. ``M.min()``.
```

stefanv commented on doc/source/reference/sparse.migration_to_sparray.rst:
```diff
@@ -0,0 +1,268 @@
+   -  scalar multiplication uses ``*`` and is not implemented for ``@``.
+   -  sparrays use ``*`` for elementwise multiplication and ``@`` for
+      matrix multiplication while spmatrices can use both operators
+      ``*`` and @ for matrix multiplication. Either can use
```
```suggestion
      ``*`` and ``@`` for matrix multiplication. Either can use
```

stefanv commented on doc/source/reference/sparse.migration_to_sparray.rst:
```diff
@@ -0,0 +1,268 @@
+   are used as 0D values where needed.
+-  Operators change: ``*, @, *=, @=, **``
+
+   -  scalar multiplication uses ``*`` and is not implemented for ``@``.
```
Which scalars are we talking about here?

stefanv commented on doc/source/reference/sparse.migration_to_sparray.rst:
```diff
@@ -0,0 +1,268 @@
+      use ``*``.
+   -  scalar powers, e.g. ``M**3``, should be converted to
+      ``sp.sparse.linalg.matrix_power(M, 3)`` or ``M @ M @ M`` if the
+      power is known.
```
Presumably the first call is faster than the second.

dschult replied:
I removed the slower multiple calls to the `@` operator to not confuse things.

stefanv commented on doc/source/reference/sparse.migration_to_sparray.rst:
```diff
@@ -0,0 +1,268 @@
+   |diags    | diags_array |
+   |spdiags  | diags_array |
+   |bmat     | block       |
+   |rand     | random_arra |
```
typos

stefanv reviewed (commented):
Quick first pass, will do another round later. For now, the most confusing aspect to me is the language around scalars, and how that is relevant for sparse matrices/arrays.

dschult committed:
add clarity suggestions [docs only]

dschult committed:
revert numpy rand changes (only intended scipy changes) [docs only]

dschult committed:
improve wording for handling isspmatrix_csr and isspmatrix [docs only]

melissawm commented:
Just a note - you can now use markdown in the scipy docs and it will render just the same as rst 😉 See, for example, https://github.com/scipy/scipy/blob/main/doc/source/tutorial/interpolate/interp_transition_guide.md

dschult committed:
add treatment of iteration when migrating

rgommers committed:
Merge branch 'main' into sparray_migration

rgommers committed:
Full copy-edit of the sparse arrays migration guide

- fix headings
- make text consistent stylistically
- fill in some missing text where there were only bullet points
- fix some omissions like `get_shape` for removed methods

rgommers committed:
Remove sparse migration guide from release.rst, it doesn't belong there

[skip actions] [skip cirrus]

rgommers commented:
This is looking pretty good! I did a full copy-edit of the migration guide. Still having a look at `sparse/__init__.py` changes, because some are off - but content-wise this is about ready to go.

rgommers commented:
A couple of quick non-content notes:
- in reST the heading levels actually matter globally, you cannot change `----` to `===` above and below a header. This won't give warnings that are caught in CI, but a lot of things will go wrong (e.g. left and right sidebars both have issues.
- for doc-only PRs, it'd be nice to use the `[docs only]` or another relevant skip command, see http://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skip-ci

rgommers committed:
fix sidebars. also move the submodules header up, and rename exceptions

[docs only]

rgommers committed:
remove unnecessary warning directive

[docs only]

rgommers commented on scipy/sparse/__init__.py:
```diff
@@ -84,53 +77,77 @@
-Building sparse matrices:
+.. _stacking-arrays:
+
+Stacking arrays
```
I wouldn't call this "stacking". Typically that term is reserved for `*stack` functions, and not for things like `kron` or `block_diag`. The existing text for this is more appropriate I think.

dschult replied:
I changed "stacking" to "combining".

rgommers commented on scipy/sparse/__init__.py:
```diff
@@ -84,53 +77,77 @@
+-------------------------
+
+- use ``isinstance(A, sp.sparse.sparray)`` to check whether an array or matrix.
+- use ``A.format == 'csr'`` to check the sparse format
```
This was already there, but ideally it shouldn't be - this belongs in a usage guide. And perhaps also in the docstring of `issparse` as a hint. This page is all function/class listings, these two bullet points are outliers (and they're not much more important than many other "how to use" docs). Let's move this out?

dschult replied:
I removed the references to non-function idioms in `__init__.py` and added them to the `issparse` doc_string.

rgommers commented on scipy/sparse/__init__.py:
```diff
@@ -84,53 +77,77 @@
+   random - Random values in a given shape matrix
+   rand - Random values in a given shape matrix (old interface)
+
+**Stacking matrices use the same functions as for** `stacking arrays`_
```
minor reST thing: we normally use dashes for reST labels, not spaces (a bit easier to work with, just like spaces in paths)

dschult replied:
Thanks for this tip. Makes sense to me.

rgommers commented on doc/source/reference/sparse.migration_to_sparray.rst:
```diff
@@ -0,0 +1,286 @@
+   |bmat     | block       |
+   |rand     | random_array|
+   |random   | random_array|
+   +=========+=============+
```
These tables don't render particularly nicely, but I'm happy to merge them - the perfect is the enemy of the good here I think.

dschult replied:
I changed the tables to rst simple table format in the latest commit. I like the ascii format better but this scales with browser size better. Let me know if you prefer to return to the ascii or suggest another way to make it work.

rgommers reviewed (commented):
A few last comments for today (nothing major).

@dschult are you happy with merging this, or did you plan to do more? I think it's ready, but the PR description still says draft.

dschult commented:
> * for doc-only PRs, it'd be nice to use the `[docs only]` or another relevant skip command, see http://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skip-ci

Can this skip command set the entire PR to be treated as `[docs only]`?  I used the command on many of the commits here, but its hard to remember to add that to the commit message every time.

dschult committed:
rename stacking

[docs only]

dschult committed:
move isinstance notes to issparse doc_string

[docs only]

dschult committed:
change table format

[docs only]

perimosocordiae commented on scipy/sparse/_base.py:
```diff
@@ -1396,6 +1396,11 @@ def issparse(x):
+    Notes
+    -----
+    Use `isinstance(x, sp.sparse.sparray)` to check between an array or matrix.
+    Use `a.format to check the sparse format, e.g. `a.format == 'csr'`.
```
```suggestion
    Use `a.format` to check the sparse format, e.g. `a.format == 'csr'`.
```

dschult committed:
fix typo

Co-authored-by: CJ Carey <perimosocordiae@gmail.com>

perimosocordiae commented on scipy/sparse/__init__.py:
```diff
@@ -10,14 +10,15 @@
    This package is switching to an array interface, compatible with
    NumPy arrays, from the older matrix interface.  We recommend that
-   you use the array objects (`bsr_array`, `coo_array`, etc.) for
+   you use the array objects (``bsr_array``, ``coo_array``, etc.) for
```
I think we still want single-backticks here, as those will produce links to the class definitions.

perimosocordiae commented on scipy/sparse/__init__.py:
```diff
@@ -26,23 +27,30 @@
 
-   The construction utilities (`eye`, `kron`, `random`, `diags`, etc.)
-   have not yet been ported, but their results can be wrapped into arrays::
+   The construction utilities (``eye``, ``kron``, ``random``, ``diags``, etc.)
```
Single-backticks here as well.

perimosocordiae commented on scipy/sparse/__init__.py:
```diff
@@ -26,23 +27,30 @@
-   The construction utilities (`eye`, `kron`, `random`, `diags`, etc.)
-   have not yet been ported, but their results can be wrapped into arrays::
+   The construction utilities (``eye``, ``kron``, ``random``, ``diags``, etc.)
+   have appropriate replacement :ref:`sparse-construction-functions`.
```
```suggestion
   have appropriate replacement (see :ref:`sparse-construction-functions`).
```

perimosocordiae commented on scipy/sparse/__init__.py:
```diff
@@ -168,59 +175,57 @@
+    7. dia_array: DIAgonal format
+
+To construct an array efficiently, use any of ``coo_array``,
+``dok_array`` or ``lil_array``. ``dok_array`` and ``lil_array``
```
Single backticks

dschult committed:
apply suggestions [docs only]
